src/mongoc/mongoc-uri.c

```diff
@@ -55,6 +55,21 @@ mongoc_uri_do_unescape (char **str)
    }
 }
 
+void
+mongoc_uri_copy_lowercase (const char *src,
+                           char *buf /* OUT */,
+                           int len)
+{
+   bson_unichar_t c;
+   const char *iter;
+   char *buf_iter;
+
+   for (iter = src, buf_iter = buf;
+        iter && *iter && (c = bson_utf8_get_char(iter)) && buf_iter - buf < len;
+        iter = bson_utf8_next_char(iter), buf_iter++) {
+      *buf_iter = tolower(c);
```

Comment:
tolower returns an int.  You'll need another strategy to move through buf_iter that isn't bytewise if the input isn't ascii

Also, unsure if migrating to lower case changes the size of the output.

... honestly, I'm not sure I know how this is supposed to work.